ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/CanHardFork.hs

```diff
@@ -73,20 +81,121 @@ import qualified Shelley.Spec.Ledger.UTxO as SL
+  where 'cpuProtocolVersion'/'cpuProtocolParameters' are the version and
+  parameters from the update. This then ends up in the following callstack
+
+  > applyChainTick
+  > |
+  > \-- epochTransition
+  >     |
+  >     \-- registerEpoch
+  >         |
+  >         \-- tryBumpVersion
+
+  Now, if this is changing the major version of the protocol, then this actually
+  indicates the transition to Shelley, and the Byron 'applyChainTick' won't
+  actually happen. Instead, in 'singleEraTransition' we will report the
+  'EpochNo' of the transition as soon as it's @2k@ (not @4k@!) deep: in other
+  words, as soon as it is stable; at this point, the HFC will do the rest.
+
+  A slightly subtle point is that the Byron ledger does not record any
+  information about /past/ updates to the protocol parameters, and so if we
+  /were/ to ask the Byron ledger /after/ the update when the transition is
+  going to take place (did take place), it will say 'Nothing': transition not
+  yet known. In practice this won't matter, as it will have been translated to
+  a Shelley ledger at that point.
+-------------------------------------------------------------------------------}
+
+byronTransition :: PartialLedgerConfig ByronBlock
+                -> EraParams
+                -> Bound
+                -> LedgerState ByronBlock
+                -> Maybe EpochNo
+byronTransition ByronPartialLedgerConfig{..}
+                EraParams{..}
+                eraStart
+                st@ByronLedgerState{..} =
+      fmap cpuEpoch
+    . latest
+    . filter isStable
+    . CC.Update.candidateProtocolUpdates
+    . CC.cvsUpdateState
+    $ byronLedgerState
+  where
+    SecurityParam k = Byron.genesisSecurityParam byronLedgerConfig
+
+    cpuSlot :: CC.Update.CandidateProtocolUpdate -> SlotNo
+    cpuSlot = Byron.fromByronSlotNo . CC.Update.cpuSlot
+
+    -- This follows the same structure as the computation in the A/B test. Let
+    -- @s@ be the slot of the update proposal. Note that the very first slot in
+    -- which the transition /could/ occur is @s + 1@; adding the required
+    -- stability, the first slot in which the transition could occur is @s + 4k
+    -- + 1@. This means that the last slot which /must/ be in /this/ era is @s +
+    -- 4k@. Hence the last /epoch/ that must be in this era is @epoch (s + 4k)@,
+    -- and the first epoch of the /next/ era is @succ (epoch (s + 4k)@.
+    cpuEpoch :: CC.Update.CandidateProtocolUpdate -> EpochNo
+    cpuEpoch upd = succ (slotToEpoch $ History.addSlots (4 * k) (cpuSlot upd))
+
+    -- Slot conversion (valid for slots in this era only)
+    slotToEpoch :: SlotNo -> EpochNo
+    slotToEpoch s =
+        History.addEpochs
+          (History.countSlots s (boundSlot eraStart) `div` unEpochSize eraEpochSize)
+          (boundEpoch eraStart)
+
+    isStable :: CC.Update.CandidateProtocolUpdate -> Bool
+    isStable upd = confirmationDepth >= 2 * k
+      where
+        confirmedInSlot :: SlotNo
```

Comment:
Isn't is "endorsed", not "confirmed"? It's totally possible I'm confused here; this stuff just slides right out of my brain every time I put it in there.

Reply:
Yup.